contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,461 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.4;
+
+import {SafeMath} from "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {AutocallVaultStorage} from "../../storage/AutocallVaultStorage.sol";
+import {
+    VaultLifecycleTreasury
+} from "../../libraries/VaultLifecycleTreasury.sol";
+import {Vault} from "../../libraries/Vault.sol";
+import {RibbonTreasuryVaultLite} from "./RibbonTreasuryVaultLite.sol";
+
+import {
+    IOtoken,
+    IController,
+    IOracle
+} from "../../interfaces/GammaInterface.sol";
+
+contract RibbonAutocallVault is RibbonTreasuryVaultLite, AutocallVaultStorage {
+    // Denominator for all pct calculations
+    uint256 internal constant PCT_MULTIPLIER = 100**2;
+
+    IOracle public immutable ORACLE;
+
+    /************************************************
+     *  EVENTS
+     ***********************************************/
+
+    event OptionTypeSet(OptionType optionType);
+
+    event CouponStateSet(
+        CouponType couponType,
+        uint256 newAutocallBarrierPCT,
+        uint256 newCouponBarrierPCT
+    );
+
+    event PeriodSet(uint256 period, uint256 newPeriod);
+
+    event ObservationPeriodFreqSet(
+        uint256 observationPeriodFreq,
+        uint256 newObservationPeriodFreq
+    );
+
+    /************************************************
+     *  CONSTRUCTOR & INITIALIZATION
+     ***********************************************/
+
+    /**
+     * @notice Initializes the contract with immutable variables
+     * @param _usdc is the USDC contract
+     * @param _oTokenFactory is the contract address for minting new opyn option types (strikes, asset, expiry)
+     * @param _gammaController is the contract address for opyn actions
+     * @param _marginPool is the contract address for providing collateral to opyn
+     */
+    constructor(
+        address _usdc,
+        address _oTokenFactory,
+        address _gammaController,
+        address _marginPool
+    )
+        RibbonTreasuryVaultLite(
+            _usdc,
+            _oTokenFactory,
+            _gammaController,
+            _marginPool
+        )
+    {
+        ORACLE = IOracle(IController(_gammaController).oracle());
+    }
+
+    /**
+     * @notice Initializes the OptionVault contract with storage variables.
+     * @param _initParams is the struct with vault initialization parameters
+     * @param _vaultParams is the struct with vault general data
+     * @param _optionType is type of the next put option
+     * @param _couponState is the coupon state
+     */
+    function initialize(
```

Comment:
similar to [this](https://github.com/ribbon-finance/ribbon-v2/pull/311/files#r1181614904)